<a href="https://colab.research.google.com/github/AaryanGotad/ml-practice/blob/main/tensorflow_practice/00_tensorflow_fundamentals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# In this notebook, we're going to cover some of the most fundamental concepts of tensors using tensorflow

More specifically, we're going to cover:
* introduction to tensors
* Getting information from tensors
* Manipulating tensors
* Tensors & NumPy
* Using @tf.function (a way to speed up your regular Python functions)
* Using GPUs with TensorFlow (or TPUs)
* Exercises to try for yourself

## Introduction to Tensors

In [ ]:
# Import TensorFlow
import tensorflow as tf
print(tf.__version__)

2.19.0


In [ ]:
# Create tensors with tf.constant()
scalar = tf.constant(7)
scalar

<tf.Tensor: shape=(), dtype=int32, numpy=7>

In [ ]:
# Check the number of dimensions of a tensor (ndim stands for number of dimensions)
scalar.ndim

0

In [ ]:
# Create a vector
vector = tf.constant([10, 10])
vector

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([10, 10], dtype=int32)>

In [ ]:
# Check the dimension of our vector
vector.ndim

1

In [ ]:
# Create a matrix (a matrix has more than one dimension)
matrix = tf.constant([[10, 7],
                      [7, 10]])
matrix

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 7, 10]], dtype=int32)>

In [ ]:
# Check the dimension of our matrix
matrix.ndim

2

In [ ]:
# Create another matrix
another_matrix = tf.constant([[10., 7.],
                              [3., 2.],
                              [8., 9.]], dtype=tf.float16) # specify the datatupe with dtype paramater

another_matrix

<tf.Tensor: shape=(3, 2), dtype=float16, numpy=
array([[10.,  7.],
       [ 3.,  2.],
       [ 8.,  9.]], dtype=float16)>

In [ ]:
# Check the dimension of another_matrix
another_matrix.ndim

2

In [ ]:
# Lets create a tensor
tensor = tf.constant([[[1, 2, 3],
                       [4, 5, 6]],
                      [[7, 8, 9],
                       [10, 11, 12]],
                      [[13, 14, 15],
                       [16, 17, 18]]])
tensor

<tf.Tensor: shape=(3, 2, 3), dtype=int32, numpy=
array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18]]], dtype=int32)>

In [ ]:
# Check the dimension of tensor
tensor.ndim

3

What we've created so far:
* Scalar: a single number
* Vector: a number with direction (e.g. wind speed and direction)
* Matrix: a 2-dimensional array of numbers
* Tensor: an n-dimensional array of numbers (where n can be any number, a 0-dimensional tensor is a scalar, a 1-dimensional tensor is a vector)

## Creating tensors with `tf.variable`

In [ ]:
tf.Variable

tensorflow.python.ops.variables.Variable

In [ ]:
# Create the same tensor with tf.Variable() as above
changeable_tensor = tf.Variable([10, 7])
unchangeable_tensor = tf.constant([10, 7])
changeable_tensor, unchangeable_tensor

(<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([10,  7], dtype=int32)>,
 <tf.Tensor: shape=(2,), dtype=int32, numpy=array([10,  7], dtype=int32)>)

In [ ]:
# Let's try changing one of the elements in our changeable_tensor
changeable_tensor[0] = 7
changeable_tensor

TypeError: 'ResourceVariable' object does not support item assignment

In [ ]:
# How about we try .assign()
changeable_tensor[0].assign(7)
changeable_tensor

<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([7, 7], dtype=int32)>

In [ ]:
# Let's try changing our unchangeable tensor
unchangeable_tensor[0] = 7
unchangeable_tensor

TypeError: 'tensorflow.python.framework.ops.EagerTensor' object does not support item assignment

In [ ]:
# Let's try .assign() here aswell
unchangeable_tensor[0].assign(7)
unchangeable_tensor

AttributeError: 'tensorflow.python.framework.ops.EagerTensor' object has no attribute 'assign'

🔑 __Note__: Rarely in practice you need to decide whether to use `tf.constanct()` or `tf.Variable()` to create tensors, as TensorFlow does this for you. However, if in doubt, use `tf.constant()` and change it later if needed.

## Creating Random Tensors

Random tensors are tensors of some arbitrary size which contain random numbers

In [ ]:
# Create two random (but the same) tensors
random_1 = tf.random.Generator.from_seed(42) # set seed for reproducibility
random_1 = random_1.normal(shape=(3, 2)) # generates random tensors from a normal distribution
random_2 = tf.random.Generator.from_seed(42)
random_2 = random_2.normal(shape=(3, 2))

# are they equal?
random_1, random_2, random_1 == random_2

(<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702],
        [ 0.07595026, -1.2573844 ],
        [-0.23193763, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702],
        [ 0.07595026, -1.2573844 ],
        [-0.23193763, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=bool, numpy=
 array([[ True,  True],
        [ True,  True],
        [ True,  True]])>)

## Shuffle the order of elements in a tensor

In [ ]:
# Shuffle a tensor (valuable for when you want to shuufle your data so the inherent order doesn't affect learning)
not_shuffled = tf.constant([[10, 7],
                            [3, 4],
                            [2, 5]])

# Shuffle our non_shuffled tensor
tf.random.shuffle(not_shuffled)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [10,  7],
       [ 3,  4]], dtype=int32)>

In [ ]:
# Shuffle our non_shuffled tensor
tf.random.set_seed(42)
tf.random.shuffle(not_shuffled, seed=42)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

In [ ]:
not_shuffled

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

🛠 Exercise: Read through TensorFlow documentation on random seed generation: https://www.tensorflow.org/api_docs/python/tf/random/set_seed and practice writing 5 random tensors and shuffle them.

It looks like if we want our shuffled tensors to be in the same order, we've got to use the global level random seed as well as the operation level random seed:

Rule 4: "If both the global and the operation seed are set: Both seeds are used in conjunction to determine the random sequence."

## Exercise

There are 4 ways random operations are performed.

1. Neither the global nor the operation seed is given
2. Global seed is set but operation seed is not
3. Operation seed is set but global seed is not
4. Lastly, both glovbal and operation seed are set

#### First, lets look what will happen if neither global seed nor the operation seed is set

In this case we get different results for every call to the random op **AND** every re-run of the program.

In [ ]:
print(tf.random.uniform([1]))
print(tf.random.uniform([1]))

tf.Tensor([0.6645621], shape=(1,), dtype=float32)
tf.Tensor([0.68789124], shape=(1,), dtype=float32)


_(Now closing the program and re running it)_

In [ ]:
print(tf.random.uniform([1]))
print(tf.random.uniform([1]))

tf.Tensor([0.7413678], shape=(1,), dtype=float32)
tf.Tensor([0.7402308], shape=(1,), dtype=float32)


### Second, if the global seed is set but the operation seed is not

Here we get different results for every call to the random op, but the sequence would stay the same for every re-run of the program

In [ ]:
tf.random.set_seed(1234)
print(tf.random.uniform([1]))
print(tf.random.uniform([1]))

tf.Tensor([0.5380393], shape=(1,), dtype=float32)
tf.Tensor([0.3253647], shape=(1,), dtype=float32)


_(Now closing the program and re running it)_

In [ ]:
tf.random.set_seed(1234)
print(tf.random.uniform([1]))
print(tf.random.uniform([1]))

tf.Tensor([0.5380393], shape=(1,), dtype=float32)
tf.Tensor([0.3253647], shape=(1,), dtype=float32)


_The reason we get_ "[0.3253647]" _instead of_ "[0.5380393]" _on the second call of `tf.random.uniform` above is because the second call uses a different operation seed._

### Third, if the operation seed is set but the global seed is not

We get different results for every call to the random op, but the same sequence for every re-run of the program.

_(Similiar to the second case)_

In [ ]:
print(tf.random.uniform([1], seed=1))
print(tf.random.uniform([1], seed=1))

tf.Tensor([0.1689806], shape=(1,), dtype=float32)
tf.Tensor([0.7539084], shape=(1,), dtype=float32)


_(Now closing the program and re-running it)_

In [ ]:
print(tf.random.uniform([1], seed=1234))
print(tf.random.uniform([1], seed=1234))

tf.Tensor([0.73822105], shape=(1,), dtype=float32)
tf.Tensor([0.20240712], shape=(1,), dtype=float32)


### Fourth, the last case when both the global and operation seed are set

Here are two cases,
* The global seed and operation seed are different-
  * Here the results are the same as the third and second case
* The global seed and operation seed are same-
  * Here's where things change, individual random op this provide random results, but the sequence is same, even if we close the program and run it again

In [ ]:
tf.random.set_seed(1234)
print(tf.random.uniform([1], seed=1234))
print(tf.random.uniform([1], seed=1234))

tf.Tensor([0.73822105], shape=(1,), dtype=float32)
tf.Tensor([0.20240712], shape=(1,), dtype=float32)


In [ ]:
tf.random.set_seed(1234)
print(tf.random.uniform([1], seed=1234))
print(tf.random.uniform([1], seed=1234))

tf.Tensor([0.73822105], shape=(1,), dtype=float32)
tf.Tensor([0.20240712], shape=(1,), dtype=float32)


## Shuffling of tensors
Now using all the knowledge we gained about random operations and how global and operation seeds work, lets shuffle some tensors with it shall we!

#### Case 1, no global and operation seeds set

In [ ]:
not_shuffled = tf.constant([[10, 7],
                            [3, 4],
                            [2, 5]])
tf.random.shuffle(not_shuffled)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [10,  7],
       [ 3,  4]], dtype=int32)>

_(Closing and re-running the program)_

In [ ]:
tf.random.shuffle(not_shuffled)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 3,  4],
       [ 2,  5],
       [10,  7]], dtype=int32)>

As, you can see the results are not the same across to runs of the program

#### Case 2, Global seed set but operation seed is not

In [ ]:
tf.random.set_seed(1234)
tf.random.shuffle(not_shuffled)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [10,  7],
       [ 3,  4]], dtype=int32)>

In [ ]:
tf.random.set_seed(1234)
tf.random.shuffle(not_shuffled)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [10,  7],
       [ 3,  4]], dtype=int32)>

Here, the shuffle is same across re-runs of the program

#### Case 3, operation seed is set but global seed is not

In [ ]:
tf.random.shuffle(not_shuffled, seed=12)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

_(Closing and re running the program)_

In [ ]:
tf.random.shuffle(not_shuffled, seed=12)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 3,  4],
       [10,  7],
       [ 2,  5]], dtype=int32)>

The results are the same as the previous case

#### Case 4, Now lets see the difference in setting both the global and operation seed

In [ ]:
tf.random.set_seed(12)
tf.random.shuffle(not_shuffled, seed=12)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [ 3,  4],
       [10,  7]], dtype=int32)>

_(Closing the program and re running it)_

In [ ]:
tf.random.set_seed(12)
tf.random.shuffle(not_shuffled, seed=12)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [ 3,  4],
       [10,  7]], dtype=int32)>

In this case the order of shuffling stays same as well as the randomness of the individual random op, if there is a difference in either of the two seeds then only the order would be the same but not the individual random operation

## Other ways to make tensors

In [ ]:
# Creating a tensor of all ones
tf.ones([10, 7])

<tf.Tensor: shape=(10, 7), dtype=float32, numpy=
array([[1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.]], dtype=float32)>

In [ ]:
# Creating a tensor of all zeroes
tf.zeros(shape=(3, 4))

<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)>

### Turning NumPy arrays into tensors

The main difference between NumPy arrays and TensorFlow tensors is that tensors can be run on a GPU (much faster for numerical computing).

In [ ]:
import numpy as np
numpy_A = np.arange(1, 25, dtype=np.int32) # creates a numpy array between 1 and 25
numpy_A
# X = tf.constant(some_matrix) # capital for matrix or tensor
# y = tf.constant(vector) # non-capital for vector

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)

In [ ]:
A = tf.constant(numpy_A)
A

<tf.Tensor: shape=(24,), dtype=int32, numpy=
array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)>

In [ ]:
A = tf.constant(numpy_A)
B = tf.constant(numpy_A, shape=(2, 3, 4))
A, B

(<tf.Tensor: shape=(24,), dtype=int32, numpy=
 array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23, 24], dtype=int32)>,
 <tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
 array([[[ 1,  2,  3,  4],
         [ 5,  6,  7,  8],
         [ 9, 10, 11, 12]],
 
        [[13, 14, 15, 16],
         [17, 18, 19, 20],
         [21, 22, 23, 24]]], dtype=int32)>)

In [ ]:
A.ndim

1

In [ ]:
B.ndim

3

### Getting information from tensors

When dealing with tensors we probably want to be aware of the following attributes:
* Shape: The length (number of elements) of each of the dimensions of a tensor.
  * Code: `tensor.shape`
* Rank: The number of tensor dimensions. A scalar has rank 0, a vector has rank 1, a matrix is rank 2, a tensor has rank n.
  * Code: `tensor.ndim`
* Axis or dimension: A particular dimension of a tensor.
  * Code: `tensor[0]`, `tensor[:, 1]`
* Size: The total number of items in the tensor.
  * Code: `tf.size(tensor)`

In [ ]:
# Create a rank 4 tensor (4 dimensions)
rank_4_tensor = tf.zeros(shape=[2, 3, 4, 5])
rank_4_tensor

<tf.Tensor: shape=(2, 3, 4, 5), dtype=float32, numpy=
array([[[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]],


       [[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]]], dtype=float32)>

In [ ]:
rank_4_tensor[0]

<tf.Tensor: shape=(3, 4, 5), dtype=float32, numpy=
array([[[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]]], dtype=float32)>

In [ ]:
rank_4_tensor.shape, rank_4_tensor.ndim, tf.size(rank_4_tensor)

(TensorShape([2, 3, 4, 5]), 4, <tf.Tensor: shape=(), dtype=int32, numpy=120>)

In [ ]:
2 * 3 * 4 * 5

120

In [ ]:
# Get various attributes of our tensor
print("Datatype of every elements:", rank_4_tensor.dtype)
print("Number of dimensions (rank):", rank_4_tensor.ndim)
print("Shape of tensor:", rank_4_tensor.shape)
print("Elements along zero axis:", rank_4_tensor.shape[0])
print("Elements along the last axis:", rank_4_tensor.shape[-1])
print("Total number of elements in our tensor:", tf.size(rank_4_tensor))
print("Total number of elements in our tensor:", tf.size(rank_4_tensor).numpy())

Datatype of every elements: <dtype: 'float32'>
Number of dimensions (rank): 4
Shape of tensor: (2, 3, 4, 5)
Elements along zero axis: 2
Elements along the last axis: 5
Total number of elements in our tensor: tf.Tensor(120, shape=(), dtype=int32)
Total number of elements in our tensor: 120


### Indexing tensors

Tensors can be indexed just like Python lists.

In [ ]:
some_list = [1, 2, 3, 4]
some_list[:2]

[1, 2]

In [ ]:
# Get the first two elements of each dimension
rank_4_tensor[:2, :2, :2, :2]

<tf.Tensor: shape=(2, 2, 2, 2), dtype=float32, numpy=
array([[[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]],


       [[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]]], dtype=float32)>

In [ ]:
some_list[:1]

[1]

In [ ]:
# Get the first element from each dimension from each index except for the final one
rank_4_tensor[:1, :1, :1]

<tf.Tensor: shape=(1, 1, 1, 5), dtype=float32, numpy=array([[[[0., 0., 0., 0., 0.]]]], dtype=float32)>

In [ ]:
rank_4_tensor[:1, :1, :1, :]

<tf.Tensor: shape=(1, 1, 1, 5), dtype=float32, numpy=array([[[[0., 0., 0., 0., 0.]]]], dtype=float32)>

In [ ]:
rank_4_tensor[:1, :1, :, :1]

<tf.Tensor: shape=(1, 1, 4, 1), dtype=float32, numpy=
array([[[[0.],
         [0.],
         [0.],
         [0.]]]], dtype=float32)>

In [ ]:
rank_4_tensor[:1, :, :1, :1]

<tf.Tensor: shape=(1, 3, 1, 1), dtype=float32, numpy=
array([[[[0.]],

        [[0.]],

        [[0.]]]], dtype=float32)>

In [ ]:
rank_4_tensor[:, :1, :1, :1]

<tf.Tensor: shape=(2, 1, 1, 1), dtype=float32, numpy=
array([[[[0.]]],


       [[[0.]]]], dtype=float32)>

In [ ]:
# Create a rank 2 tensor (2 dimensions)
rank_2_tensor = tf.constant([[10, 7],
                             [3, 4]])
rank_2_tensor.shape, rank_2_tensor.ndim

(TensorShape([2, 2]), 2)

In [ ]:
rank_2_tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

In [ ]:
some_list_2 = [[10, 7],
               [3, 4]]
some_list_2[-1]

[3, 4]

In [ ]:
# Get the last item of our rank 2 tensor
rank_2_tensor[:, -1]

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([7, 4], dtype=int32)>

In [ ]:
# Add in extra dimension to our tank 2 tensor
rank_3_tensor = rank_2_tensor[..., tf.newaxis] # '...' is a replacemnt for ':, :, :', basically saying include every axis before the last one and add a new axis at the end
rank_3_tensor

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]], dtype=int32)>

In [ ]:
# Alternatives to tf.newaxis
tf.expand_dims(rank_2_tensor, axis=-1) # '-1' means expand the final axis

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]], dtype=int32)>

In [ ]:
# Expand the 0-axis
tf.expand_dims(rank_2_tensor, axis=0) # expand the 0-axis

<tf.Tensor: shape=(1, 2, 2), dtype=int32, numpy=
array([[[10,  7],
        [ 3,  4]]], dtype=int32)>

In [ ]:
rank_2_tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

### Manipulating tensors (Tensor operations)

**Basic Operations**

`+`, `-`, `*`, `/`

In [ ]:
# You can add values to a tensor using the addition operator
tensor = tf.constant([[10, 7],
                      [3, 4]])
tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

In [ ]:
# Original tenosr is unchanged
tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

In [ ]:
# Multiplication also works
tensor * 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[100,  70],
       [ 30,  40]], dtype=int32)>

In [ ]:
# Subtraction if you want
tensor - 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 0, -3],
       [-7, -6]], dtype=int32)>

In [ ]:
# We can use the tensorflow built-in function too
tf.multiply(tensor, 10) # this is faster than traditional operations we did just now

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[100,  70],
       [ 30,  40]], dtype=int32)>

In [ ]:
tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

## **Matrix Multiplication**

In machine learning, matrix multiplication is one of the most common tensor operations

In [ ]:
# Matrix multiplication in tensorflow
print(tensor)

tf.Tensor(
[[10  7]
 [ 3  4]], shape=(2, 2), dtype=int32)


In [ ]:
tf.linalg.matmul(tensor, tensor)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]], dtype=int32)>

In [ ]:
tf.matmul(tensor, tensor) # we can generally drop the middle package name

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]], dtype=int32)>

In [ ]:
tensor * tensor # this gives a diff. output as this is element wise multiplication rather than matrix multiplication

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[100,  49],
       [  9,  16]], dtype=int32)>

### Excercise of matrix.xyz website in lecture

matrix 1 = [ 1, 2, 5<br>
           &emsp;&emsp;&emsp;&emsp;&ensp;&nbsp;&nbsp;&nbsp;7, 2, 1<br>
           &emsp;&emsp;&emsp;&emsp;&ensp;&nbsp;&nbsp;&nbsp;3, 3, 3 ]<br>

matrix 2 = [ 3, 5<br>
           &emsp;&emsp;&emsp;&emsp;&ensp;&nbsp;&nbsp;&nbsp;6, 7<br>
           &emsp;&emsp;&emsp;&emsp;&ensp;&nbsp;&nbsp;&nbsp;1, 8 ]<br>

matrix 1 * matrix 2 should be equal to = [ 20, 59<br>
     &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;&ensp;&nbsp;&nbsp;&nbsp;34, 57<br>                        &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;&ensp;&nbsp;&nbsp;&nbsp;30, 60 ]<br>

In [ ]:
matrix_1 = tf.constant([[1, 2, 5],
                        [7, 2, 1],
                        [3, 3, 3]])
matrix_2 = tf.constant([[3, 5],
                        [6, 7],
                        [1, 8]])
multiplication = tf.matmul(matrix_1, matrix_2)
multiplication

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[20, 59],
       [34, 57],
       [30, 60]], dtype=int32)>

In [ ]:
# Matrix multiplication with Python operator "@"
tensor @ tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]], dtype=int32)>

In [ ]:
tensor.shape

TensorShape([2, 2])

## **The Dot Product**

Matrix multiplication is also referred to as the dot product
You can perform matrix multiplication using:
* `tf.matmul()`
* `tf.tensordot()`
* `@`

In [ ]:
X = tf.constant([[1, 2],
                 [3, 4],
                 [5, 6]])
Y = tf.constant([[7, 8],
                 [9, 10],
                 [11, 12]])

X, Y

(<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
 array([[1, 2],
        [3, 4],
        [5, 6]], dtype=int32)>,
 <tf.Tensor: shape=(3, 2), dtype=int32, numpy=
 array([[ 7,  8],
        [ 9, 10],
        [11, 12]], dtype=int32)>)

In [ ]:
# Perform the dot product on X and Y (transposed)
tf.tensordot(tf.transpose(X), Y, axes=1)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 89,  98],
       [116, 128]], dtype=int32)>

In [ ]:
tf.tensordot(tf.reshape(X, shape=(2, 3)), Y, axes=1)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 58,  64],
       [139, 154]], dtype=int32)>

In [ ]:
# Perform matrix multiplication between X and Y (reshaped)
tf.matmul(X, tf.reshape(Y, shape=(2, 3)))

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 27,  30,  33],
       [ 61,  68,  75],
       [ 95, 106, 117]], dtype=int32)>

In [ ]:
tf.matmul(X, tf.transpose(Y))

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 23,  29,  35],
       [ 53,  67,  81],
       [ 83, 105, 127]], dtype=int32)>

In [ ]:
# Check the values of Y, reshape Y and transposed Y
print("Normal Y:")
print(Y, "\n")

print("Y reshaped to (2, 3):")
print(tf.reshape(Y, (2, 3)), "\n")

print("Y transpose:")
print(tf.transpose(Y), "\n")

Normal Y:
tf.Tensor(
[[ 7  8]
 [ 9 10]
 [11 12]], shape=(3, 2), dtype=int32) 

Y reshaped to (2, 3):
tf.Tensor(
[[ 7  8  9]
 [10 11 12]], shape=(2, 3), dtype=int32) 

Y transpose:
tf.Tensor(
[[ 7  9 11]
 [ 8 10 12]], shape=(2, 3), dtype=int32) 



Generally, when performing matrix multiplication and one of the axes doesn't line up, you will transpose rather than reshape one of the tensors to satisfy the matrix multiplication rules.  

### Changing the datatype of tensors

In [ ]:
print(tf.__version__)

2.19.0


In [ ]:
# Create a new tensor with default datatype (float32)
B = tf.constant([1.7, 7.4])
B.dtype

tf.float32

In [ ]:
C = tf.constant([7, 10])
C.dtype

tf.int32

In [ ]:
# Change from float32 to float16 (reduced precision)
D = tf.cast(B, dtype=tf.float16)
D, B.dtype

(<tf.Tensor: shape=(2,), dtype=float16, numpy=array([1.7, 7.4], dtype=float16)>,
 tf.float32)

In [ ]:
# Change from int32 to float32
E = tf.cast(C, dtype=tf.float32)
E, C.dtype

(<tf.Tensor: shape=(2,), dtype=float32, numpy=array([ 7., 10.], dtype=float32)>,
 tf.int32)

In [ ]:
E_float16 = tf.cast(E, dtype=tf.float16)
E_float16, E.dtype, C.dtype

(<tf.Tensor: shape=(2,), dtype=float16, numpy=array([ 7., 10.], dtype=float16)>,
 tf.float32,
 tf.int32)

### Aggregating tensors

Aggregating tensors = condensing them from multiple values down to a smaller amount of values

In [ ]:
# Get the absolute values
D = tf.constant([-7, -10])
tf.abs(D)

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ 7, 10], dtype=int32)>

let's go through the following forms of aggregration
* Get the tf.minimum
* Get the maximum
* Get the mean of a tensor
* Get the sum of a tensor

In [ ]:
# Create a random tensor with values between 0 and 100 of size 50
E = tf.constant(np.random.randint(1, 100, size=50))
E

<tf.Tensor: shape=(50,), dtype=int64, numpy=
array([90, 59, 28,  4, 55, 97, 15, 69, 64,  4, 34,  3, 65, 43, 35, 18, 96,
       50, 88, 63, 62, 91, 14, 99, 52, 81, 97, 34, 38, 48, 74, 18, 29, 66,
       73, 35, 64, 95, 86, 34,  1, 91, 37, 96, 54, 80, 48, 55, 61, 57])>

In [ ]:
# Find the minimum
tf.reduce_min(E)

<tf.Tensor: shape=(), dtype=int64, numpy=1>

In [ ]:
# Find the maximum
tf.reduce_max(E)

<tf.Tensor: shape=(), dtype=int64, numpy=99>

In [ ]:
tf.size(E), E.shape, E.ndim

(<tf.Tensor: shape=(), dtype=int32, numpy=50>, TensorShape([50]), 1)

In [ ]:
# Find the mean
tf.reduce_mean(E)

<tf.Tensor: shape=(), dtype=int64, numpy=55>

In [ ]:
# Find the sum
tf.reduce_sum(E)

<tf.Tensor: shape=(), dtype=int64, numpy=2750>

⚒️**Exercise** With what we've just learned, find the variance and standard deviation of our "E" tensor using Tensorflow methods.

In [ ]:
# Finding the variance
tf.math.reduce_variance(tf.cast(E, dtype=tf.float32))

<tf.Tensor: shape=(), dtype=float32, numpy=805.719970703125>

In [ ]:
# Finding standard deviation
tf.math.reduce_std(tf.cast(E, dtype=tf.float32))

<tf.Tensor: shape=(), dtype=float32, numpy=28.38520622253418>

In [ ]:
# Different way to find the standard deviation of a tensor
tf.math.sqrt(tf.math.reduce_variance(tf.cast(E, dtype=tf.float32)))

<tf.Tensor: shape=(), dtype=float32, numpy=28.38520622253418>

## **Find the positional maximum and minimum of a tensor**

Positional maximum/minimum means at what position we get the maximum/minimum value respectively of the tensor

In [ ]:
# Creating a new tensor for finding positional maximum and minimum
F = tf.random.uniform(shape=[50])
F

<tf.Tensor: shape=(50,), dtype=float32, numpy=
array([0.47720265, 0.33357763, 0.4960823 , 0.48351312, 0.2963959 ,
       0.69444907, 0.68488   , 0.6662284 , 0.07551682, 0.06608915,
       0.91428673, 0.21982336, 0.066468  , 0.37987065, 0.88209164,
       0.73124087, 0.59181285, 0.9785874 , 0.0937705 , 0.37701082,
       0.2055552 , 0.2800591 , 0.44172525, 0.9199766 , 0.49873018,
       0.10266078, 0.12177575, 0.10520744, 0.22547913, 0.39156199,
       0.5587522 , 0.7095549 , 0.04214132, 0.03028095, 0.14591146,
       0.7897457 , 0.52633774, 0.8487345 , 0.08758485, 0.5727763 ,
       0.74252594, 0.63061965, 0.03749514, 0.0432719 , 0.13088238,
       0.6540686 , 0.9438777 , 0.65461504, 0.10520577, 0.8353534 ],
      dtype=float32)>

In [ ]:
# Finding the positional maximum
tf.argmax(F)

<tf.Tensor: shape=(), dtype=int64, numpy=17>

In [ ]:
# Indexing on our largest value position
F[tf.argmax(F)]

<tf.Tensor: shape=(), dtype=float32, numpy=0.9785873889923096>

In [ ]:
# Finding the max value of F
tf.reduce_max(F)

<tf.Tensor: shape=(), dtype=float32, numpy=0.9785873889923096>

In [ ]:
# Checking for equality (Just to be sure)
F[tf.argmax(F)] == tf.reduce_max(F)

<tf.Tensor: shape=(), dtype=bool, numpy=True>

In [ ]:
# Finding the positional minimum
tf.argmin(F)

<tf.Tensor: shape=(), dtype=int64, numpy=33>

In [ ]:
F[tf.argmin(F)]

<tf.Tensor: shape=(), dtype=float32, numpy=0.0302809476852417>

In [ ]:
# Finding the min value of our tensor
tf.reduce_min(F)

<tf.Tensor: shape=(), dtype=float32, numpy=0.0302809476852417>

In [ ]:
# Checking for equality
F[tf.argmin(F)] == tf.reduce_min(F)

<tf.Tensor: shape=(), dtype=bool, numpy=True>

### Squeezing a tensor (removing all single dimensions)

In [ ]:
# Creating a tensor to get started
tf.random.set_seed(42)
G = tf.constant(tf.random.uniform(shape=[50]), shape=(1, 1, 1, 1, 50))
G

<tf.Tensor: shape=(1, 1, 1, 1, 50), dtype=float32, numpy=
array([[[[[0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
           0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
           0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
           0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
           0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ,
           0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
           0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
           0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
           0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
           0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043]]]]],
      dtype=float32)>

In [ ]:
G_squeezed = tf.squeeze(G)
G_squeezed, G_squeezed.shape

(<tf.Tensor: shape=(50,), dtype=float32, numpy=
 array([0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
        0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
        0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
        0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
        0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ,
        0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
        0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
        0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
        0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
        0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043],
       dtype=float32)>,
 TensorShape([50]))

### **One-Hot Encoding**

In [ ]:
# Create a list of indices
some_list = [0, 1, 2, 3] # could be red, green blue, purple

# One hot encode our list of indices
tf.one_hot(some_list, depth=4)

<tf.Tensor: shape=(4, 4), dtype=float32, numpy=
array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]], dtype=float32)>

In [ ]:
# Specify custom values for one hot encoding
tf.one_hot(some_list, depth=4, on_value="yo I love deep learning", off_value="I also like to dance")

<tf.Tensor: shape=(4, 4), dtype=string, numpy=
array([[b'yo I love deep learning', b'I also like to dance',
        b'I also like to dance', b'I also like to dance'],
       [b'I also like to dance', b'yo I love deep learning',
        b'I also like to dance', b'I also like to dance'],
       [b'I also like to dance', b'I also like to dance',
        b'yo I love deep learning', b'I also like to dance'],
       [b'I also like to dance', b'I also like to dance',
        b'I also like to dance', b'yo I love deep learning']],
      dtype=object)>

### Squaring, log, square root

In [ ]:
# Create a new tensor
H = tf.range(1, 10)
H

<tf.Tensor: shape=(9,), dtype=int32, numpy=array([1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=int32)>

In [ ]:
# Square it
tf.square(H)

<tf.Tensor: shape=(9,), dtype=int32, numpy=array([ 1,  4,  9, 16, 25, 36, 49, 64, 81], dtype=int32)>

In [ ]:
# Find the square root (will error, method requires non-int type)
tf.sqrt(H)

InvalidArgumentError: Value for attr 'T' of int32 is not in the list of allowed values: bfloat16, half, float, double, complex64, complex128
	; NodeDef: {{node Sqrt}}; Op<name=Sqrt; signature=x:T -> y:T; attr=T:type,allowed=[DT_BFLOAT16, DT_HALF, DT_FLOAT, DT_DOUBLE, DT_COMPLEX64, DT_COMPLEX128]> [Op:Sqrt] name: 

In [ ]:
# Find the square root
tf.sqrt(tf.cast(H, dtype=tf.float32))

<tf.Tensor: shape=(9,), dtype=float32, numpy=
array([1.       , 1.4142135, 1.7320508, 2.       , 2.236068 , 2.4494898,
       2.6457512, 2.828427 , 3.       ], dtype=float32)>

In [ ]:
# Find the log (again, will throw an error, it requires a non-int value)
tf.math.log(H)

InvalidArgumentError: Value for attr 'T' of int32 is not in the list of allowed values: bfloat16, half, float, double, complex64, complex128
	; NodeDef: {{node Log}}; Op<name=Log; signature=x:T -> y:T; attr=T:type,allowed=[DT_BFLOAT16, DT_HALF, DT_FLOAT, DT_DOUBLE, DT_COMPLEX64, DT_COMPLEX128]> [Op:Log] name: 

In [ ]:
# Find the log
tf.math.log(tf.cast(H, dtype=tf.float32))

<tf.Tensor: shape=(9,), dtype=float32, numpy=
array([0.       , 0.6931472, 1.0986123, 1.3862944, 1.609438 , 1.7917595,
       1.9459102, 2.0794415, 2.1972246], dtype=float32)>

### Tensors and NumPy

Tensorflow interacts beautifully with NumPy arrays

🔑**Note**: One of the main differences between a TensorFlow tensor and a NumPy array is that a TensorFlow tensor can be run on a GPU or TPU (for faster numerical processing)

In [ ]:
# Create a tensor directly from a NumPy array
J = tf.constant(np.array([3., 7., 10.]))
J

<tf.Tensor: shape=(3,), dtype=float64, numpy=array([ 3.,  7., 10.])>

In [ ]:
# Convert our tensor back to NumPy array
np.array(J), type(np.array(J))

(array([ 3.,  7., 10.]), numpy.ndarray)

In [ ]:
# Convert tensor J to a NumPy array
J.numpy(), type(J.numpy())

(array([ 3.,  7., 10.]), numpy.ndarray)

In [ ]:
# Default types of each are slightly different
numpy_J = tf.constant(np.array([3., 7., 10.]))
tensor_J = tf.constant([3., 7., 10.])
# Check the datatypes of each
numpy_J.dtype, tensor_J.dtype

(tf.float64, tf.float32)

### Finding access to GPUs

In [ ]:
import tensorflow as tf
tf.config.list_physical_devices('GPU')

[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

In [ ]:
!nvidia-smi

Wed Nov  5 16:31:31 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   35C    P8              9W /   70W |       2MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

🔑**Note:** if you have access to a CUDA-enabled GPU, TensorFlow will automatically use it whenever possible

# ⚒️ Excercises

## 1. Create a vector, scalar, matrix and tensor with values of your choosing using `tf.constant()`.

In [ ]:
import tensorflow as tf

# creating vector
vector = tf.constant([1, 2, 3, 4, 5])

# creating scalar
scalar = tf.constant(6)

# creating matrix
matrix = tf.constant([[7, 8, 9],
                      [10, 11, 12],
                      [13, 14, 15]])

# creating tensorf
tensor = tf.constant([
                         [
                              [16, 17, 18],
                              [19, 20, 21],
                              [22, 23, 24]
                         ],

                         [
                             [25, 26, 27],
                             [28, 29, 30],
                             [31, 32, 33]
                         ],

                         [
                             [33, 34, 35],
                             [36, 37, 38],
                             [39, 40, 41]
                         ]
                      ])

## 2. Find the shape, rank and size of the tensors you created in 1.

In [ ]:
# shapes of each tensor
scalar.shape, vector.shape, matrix.shape, tensor.shape

(TensorShape([]),
 TensorShape([5]),
 TensorShape([3, 3]),
 TensorShape([3, 3, 3]))

In [ ]:
# rank of each tensor
tf.rank(scalar), tf.rank(vector), tf.rank(matrix), tf.rank(tensor)

(<tf.Tensor: shape=(), dtype=int32, numpy=0>,
 <tf.Tensor: shape=(), dtype=int32, numpy=1>,
 <tf.Tensor: shape=(), dtype=int32, numpy=2>,
 <tf.Tensor: shape=(), dtype=int32, numpy=3>)

In [ ]:
# size of each tensor
tf.size(scalar), tf.size(vector), tf.size(matrix), tf.size(tensor)

(<tf.Tensor: shape=(), dtype=int32, numpy=1>,
 <tf.Tensor: shape=(), dtype=int32, numpy=5>,
 <tf.Tensor: shape=(), dtype=int32, numpy=9>,
 <tf.Tensor: shape=(), dtype=int32, numpy=27>)

## 3. Create two tensors containing random values between 0 and 1 with shape [5, 300].

In [ ]:
a = tf.random.uniform(shape=[5, 500])
b = tf.random.uniform(shape=[5, 500])
a, b

(<tf.Tensor: shape=(5, 500), dtype=float32, numpy=
 array([[0.690156  , 0.22461498, 0.3061583 , ..., 0.6259066 , 0.2268821 ,
         0.5851368 ],
        [0.08638895, 0.0323199 , 0.79686904, ..., 0.903046  , 0.61999416,
         0.50696313],
        [0.6306584 , 0.9070264 , 0.02705526, ..., 0.6422514 , 0.33671916,
         0.75630474],
        [0.21683705, 0.71557474, 0.64415765, ..., 0.74236166, 0.35462248,
         0.78949547],
        [0.08522391, 0.28978765, 0.21462774, ..., 0.47058976, 0.05509186,
         0.8823495 ]], dtype=float32)>,
 <tf.Tensor: shape=(5, 500), dtype=float32, numpy=
 array([[0.82948625, 0.43526185, 0.40993094, ..., 0.9643258 , 0.42547643,
         0.4173602 ],
        [0.5868852 , 0.07778645, 0.11031222, ..., 0.15739846, 0.16535413,
         0.75449204],
        [0.6572436 , 0.2857567 , 0.10375273, ..., 0.6180295 , 0.36337566,
         0.9990624 ],
        [0.9724076 , 0.2484777 , 0.56328523, ..., 0.6032772 , 0.5755607 ,
         0.4788791 ],
        [0.43036

## 4. Multiply the two tensors you created in 3 using matrix multiplication.

In [ ]:
tf.matmul(a, tf.transpose(b))

<tf.Tensor: shape=(5, 5), dtype=float32, numpy=
array([[115.97444 , 115.21391 , 125.78518 , 121.54524 , 115.74089 ],
       [114.99102 , 110.17767 , 119.08266 , 118.172516, 116.623505],
       [114.771126, 112.726364, 121.24358 , 118.44238 , 116.18369 ],
       [116.76988 , 120.45514 , 128.05539 , 122.05684 , 120.03531 ],
       [119.22205 , 115.601456, 127.06719 , 124.4458  , 117.74687 ]],
      dtype=float32)>

## 5. Multiply the two tensors you created in 3 using dot product.

In [ ]:
tf.tensordot(a, tf.transpose(b), axes=1)

<tf.Tensor: shape=(5, 5), dtype=float32, numpy=
array([[115.97444 , 115.21391 , 125.78518 , 121.54524 , 115.74089 ],
       [114.99102 , 110.17767 , 119.08266 , 118.172516, 116.623505],
       [114.771126, 112.726364, 121.24358 , 118.44238 , 116.18369 ],
       [116.76988 , 120.45514 , 128.05539 , 122.05684 , 120.03531 ],
       [119.22205 , 115.601456, 127.06719 , 124.4458  , 117.74687 ]],
      dtype=float32)>

## 6. Create a tensor with random values between 0 and 1 with shape [224, 224, 3].

In [ ]:
c = tf.random.uniform(shape=[224, 224, 3])
c

<tf.Tensor: shape=(224, 224, 3), dtype=float32, numpy=
array([[[6.9424605e-01, 9.3080568e-01, 6.1287796e-01],
        [3.8348782e-01, 6.9526005e-01, 4.0412402e-01],
        [4.1960144e-01, 2.4777186e-01, 1.1025667e-02],
        ...,
        [4.6133375e-01, 9.8774648e-01, 5.2235675e-01],
        [9.0907454e-01, 7.7636230e-01, 1.7892778e-01],
        [8.9520097e-01, 9.4775164e-01, 4.3146372e-02]],

       [[6.6270685e-01, 6.2009919e-01, 2.7836597e-01],
        [9.0459061e-01, 2.1137035e-01, 3.3738124e-01],
        [6.8242574e-01, 9.6949816e-01, 7.8234899e-01],
        ...,
        [6.5314770e-04, 1.6087079e-01, 3.9978397e-01],
        [2.3865688e-01, 9.3644524e-01, 4.6426249e-01],
        [9.4376433e-01, 1.5641451e-02, 9.5996201e-01]],

       [[5.9961629e-01, 3.8466036e-01, 6.8424010e-01],
        [6.9969535e-02, 5.4611492e-01, 1.7672384e-01],
        [5.5422235e-01, 4.9663639e-01, 2.2329998e-01],
        ...,
        [1.4000845e-01, 7.0710075e-01, 7.9389203e-01],
        [1.2590921e-01

## 7. Find the min and max values of the tensor you created in 6 along the first axis.

In [ ]:
# minimum value of the tensor created in 6.
tf.reduce_min(c, axis=0)

<tf.Tensor: shape=(224, 3), dtype=float32, numpy=
array([[1.40929222e-03, 2.48622894e-03, 1.23879910e-02],
       [7.88867474e-03, 7.20250607e-03, 8.35657120e-05],
       [3.98504734e-03, 6.80506229e-03, 4.52911854e-03],
       [2.52366066e-04, 3.46064568e-04, 8.87036324e-04],
       [6.71029091e-04, 3.84807587e-04, 2.43854523e-03],
       [6.25574589e-03, 3.61800194e-04, 3.01563740e-03],
       [2.67338753e-03, 1.59132481e-03, 7.51221180e-03],
       [1.88481808e-03, 2.76839733e-03, 3.17692757e-04],
       [3.20792198e-04, 4.63807583e-03, 4.08852100e-03],
       [4.22084332e-03, 2.49445438e-03, 4.96339798e-03],
       [1.22643709e-02, 2.54774094e-03, 9.10210609e-03],
       [4.02092934e-04, 3.13162804e-03, 3.59046459e-03],
       [4.07087803e-03, 1.20270252e-03, 8.45694542e-03],
       [1.62792206e-03, 3.76641750e-03, 1.54030323e-03],
       [2.93886662e-03, 1.84011459e-03, 6.74009323e-04],
       [5.18381596e-03, 1.82793140e-02, 6.93178177e-03],
       [1.85132027e-04, 1.32405758e-03

In [ ]:
# maximum value(s) along the first axis
tf.reduce_max(c, axis=1)

<tf.Tensor: shape=(224, 3), dtype=float32, numpy=
array([[0.998417  , 0.99026513, 0.98780406],
       [0.9867723 , 0.9949125 , 0.9989319 ],
       [0.9959438 , 0.9899262 , 0.98960924],
       [0.99946487, 0.9925138 , 0.99380016],
       [0.9994838 , 0.9996486 , 0.99744856],
       [0.99964714, 0.9982734 , 0.9920815 ],
       [0.99943364, 0.9941833 , 0.99973786],
       [0.9964199 , 0.9998983 , 0.99720466],
       [0.9986559 , 0.99377465, 0.9999614 ],
       [0.99983597, 0.999948  , 0.9995624 ],
       [0.9968811 , 0.99951124, 0.99736154],
       [0.9970499 , 0.99358857, 0.99634314],
       [0.9919765 , 0.99853003, 0.9834285 ],
       [0.9954536 , 0.98945725, 0.9932041 ],
       [0.9955195 , 0.9936241 , 0.981035  ],
       [0.9989655 , 0.99881625, 0.99522686],
       [0.9997803 , 0.99150085, 0.99722755],
       [0.99992085, 0.9998938 , 0.9923394 ],
       [0.9996096 , 0.9994693 , 0.9967946 ],
       [0.99520373, 0.9969517 , 0.99240565],
       [0.99229276, 0.9994534 , 0.9946921 ],
     

## 8. Created a tensor with random values of shape [1, 224, 224, 3] then squeeze it to change the shape to [224, 224, 3].

In [ ]:
d = tf.random.uniform(shape=[1, 224, 224, 3])
d

<tf.Tensor: shape=(1, 224, 224, 3), dtype=float32, numpy=
array([[[[4.0268660e-02, 3.9480698e-01, 2.2059917e-01],
         [9.5491755e-01, 8.6284506e-01, 4.9958587e-01],
         [7.5641561e-01, 9.4883811e-01, 6.3069689e-01],
         ...,
         [1.4056444e-02, 2.0705175e-01, 6.0616791e-01],
         [8.8966203e-01, 8.3004212e-01, 2.8354812e-01],
         [6.6497326e-03, 3.1399596e-01, 9.5480871e-01]],

        [[1.5538585e-01, 2.2205210e-01, 3.7081230e-01],
         [5.5617285e-01, 2.9157674e-01, 1.3769162e-01],
         [1.7956042e-01, 1.4837801e-01, 1.6028631e-01],
         ...,
         [4.6159756e-01, 9.9271512e-01, 4.5002460e-01],
         [3.7140703e-01, 8.5153913e-01, 8.9105380e-01],
         [7.0243001e-02, 9.7993207e-01, 5.9331429e-01]],

        [[3.2117879e-01, 9.5344329e-01, 9.2631102e-02],
         [1.1362982e-01, 6.5018213e-01, 7.0293176e-01],
         [8.0499423e-01, 6.9108677e-01, 6.2054348e-01],
         ...,
         [2.8216493e-01, 3.4582508e-01, 8.1758142e-01],


In [ ]:
# reducing the size of the tensor
# tf.reshape(d, shape=[224, 224, 3])

# or squeeze will remove any 1-valued dimension
tf.squeeze(d)

<tf.Tensor: shape=(224, 224, 3), dtype=float32, numpy=
array([[[4.0268660e-02, 3.9480698e-01, 2.2059917e-01],
        [9.5491755e-01, 8.6284506e-01, 4.9958587e-01],
        [7.5641561e-01, 9.4883811e-01, 6.3069689e-01],
        ...,
        [1.4056444e-02, 2.0705175e-01, 6.0616791e-01],
        [8.8966203e-01, 8.3004212e-01, 2.8354812e-01],
        [6.6497326e-03, 3.1399596e-01, 9.5480871e-01]],

       [[1.5538585e-01, 2.2205210e-01, 3.7081230e-01],
        [5.5617285e-01, 2.9157674e-01, 1.3769162e-01],
        [1.7956042e-01, 1.4837801e-01, 1.6028631e-01],
        ...,
        [4.6159756e-01, 9.9271512e-01, 4.5002460e-01],
        [3.7140703e-01, 8.5153913e-01, 8.9105380e-01],
        [7.0243001e-02, 9.7993207e-01, 5.9331429e-01]],

       [[3.2117879e-01, 9.5344329e-01, 9.2631102e-02],
        [1.1362982e-01, 6.5018213e-01, 7.0293176e-01],
        [8.0499423e-01, 6.9108677e-01, 6.2054348e-01],
        ...,
        [2.8216493e-01, 3.4582508e-01, 8.1758142e-01],
        [8.0125368e-01

## 9. Create a tensor with shape [10] using your own choice of values, then find the index which has the maximum value.

In [ ]:
e = tf.random.normal(shape=[10], mean=0, stddev=3)
e

<tf.Tensor: shape=(10,), dtype=float32, numpy=
array([-2.415551  ,  2.7391272 , -1.9428288 ,  4.0236893 , -1.2081168 ,
       -4.7036104 ,  0.18450963,  4.055602  , -0.7218092 ,  0.8862026 ],
      dtype=float32)>

In [ ]:
max_index = tf.argmax(e)
max_index

<tf.Tensor: shape=(), dtype=int64, numpy=7>

## 10. One-hot encode the tensor you created in 9.

In [ ]:
# casting the tensor to int before one-hot-encoding it
tf.cast(e, dtype=tf.int32)

<tf.Tensor: shape=(10,), dtype=int32, numpy=array([-2,  2, -1,  4, -1, -4,  0,  4,  0,  0], dtype=int32)>

In [ ]:
# now we can one-hot-encode it
tf.one_hot(tf.cast(e, dtype=tf.int32), depth=10)

<tf.Tensor: shape=(10, 10), dtype=float32, numpy=
array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], dtype=float32)>